src/DataSourceProvider.vue

```diff
@@ -1,19 +1,139 @@
 <template>
-  <div>
-    <h1>Your fliplet-widget-data-source-provider widget will be added here.</h1>
-    <First-Component></First-Component>
-  </div>
+  <section>
+    <div v-if="isLoading" class="spinner-holder animated">
+      <div class="spinner-overlay">Loading...</div>
+    </div>
+    <div v-else-if="hasError" class="alert alert-danger" role="alert">
+      {{ errorMessage }}
+    </div>
+    <div v-else>
+      <DataSourceSelector
+        :currentAppDataSources="appDataSources"
+        :otherDataSources="allDataSources"
+        :selectedDataSource="selectedDataSource"
+        :changeDataSource="changeDataSource"
+        :showAll="showAll"
+        @selectedDataSourceId="selectedDataSourceId = $emit"
+        @onDataSourceCreate="createDataSource"
+        @onShowAll="(event) => { showAllDataSources(event) }"
+        @onDataSourceChange="changeDataSource = !changeDataSource"
+      >
+      </DataSourceSelector>
+    </div>
+  </section>
+
 </template>
 
 <script>
-import FirstComponent from './components/FirstComponent';
+import DataSourceSelector from './components/DataSourceSelector';
+import { getDataSources, getDataSource, createDataSource } from './services/dataSource';
 
 export default {
   data() {
-    return {};
+    return {
+      appDataSources: [],
+      allDataSources: [],
+      copyOfAllDataSources: [],
+      isLoading: true,
+      hasError: false,
+      errorMessage: '',
+      widgetData: {},
+      selectedDataSourceId: false,
+      changeDataSource: false,
+      showAll: false
+    };
   },
   components: {
-    FirstComponent
+    DataSourceSelector
+  },
+  methods: {
+    showAllDataSources: function(isChecked) {
+      this.isLoading = true;
+      this.showAll = isChecked;
+
+      if (isChecked) {
+        if (!this.copyOfAllDataSources.length) {
+          this.loadDataSources();
+          return;
+        }
+
+        this.allDataSources = [...this.copyOfAllDataSources];
+      } else {
+        this.copyOfAllDataSources = [...this.allDataSources];
+        this.allDataSources = [];
+      }
+
+      // Give VUE time to reset templates
+      setTimeout(() => {
+        this.isLoading = false;
+      }, 100);
+    },
+    loadDataSources: function(appId) {
+      this.isLoading = true;
+
+      getDataSources(appId)
```

Comment:
`loadDataSources()` is the first function that is run when the provider loads.

If a data source is already selected, it looks like the provider will load all the data sources available to the app, and then it will get the information for the selected data source. That's not what we want.

If a data source is already selected, the only information the provider needs is the information on the single data source. The `getDataSources()` function call is a waste in resources.

Reply:
@tonytlwu added such logic to the data source provider.